### PR TITLE
perf: Performance with `Promise.all()`

### PR DESCRIPTION
**Pull request by sanjaiyan-dev**
Hello, this PR introduces some minor performance gains by running the asynchronous functions within `Promise.all`. This approach offers several additional benefits, including:

1. **Improved Parallelism:** By using `Promise.all` to execute asynchronous functions concurrently, we can take full advantage of the available resources and maximize parallelism, potentially speeding up the overall execution time.

2. **Reduced Latency:** When dealing with multiple asynchronous tasks, executing them concurrently can help reduce the total latency, as tasks can proceed independently instead of waiting for each other.

3. **Potential for Resource Optimization:** Running asynchronous tasks concurrently can also lead to better resource utilization, as it minimizes idle time and keeps the CPU or other resources busy with meaningful work.

_Extremely sorry if I made any mistakes :(_

### CODE REVIEW

File: fs/copy.ts [hunk 1]
```diff
@@ -190,17 +190,21 @@ async function copyDir(
   src = toPathString(src);
   dest = toPathString(dest);
 
+  let promisesCopyAsyncFunctions: Promise<void>[] = [];
```
Comment (aapoalas) [inline fs/copy.ts:193]: **nitpick:** Use `const`

File: fs/empty_dir.ts [hunk 2]
```diff
@@ -23,13 +23,14 @@ export async function emptyDir(dir: string | URL) {
...
+    await Promise.all(items.map((item) => {
       if (item && item.name) {
         const filepath = join(toPathString(dir), item.name);
-        await Deno.remove(filepath, { recursive: true });
+        return Deno.remove(filepath, { recursive: true });
       }
-    }
+    }));
+
+    items.length = 0;
```
Comment (kt3k) [inline fs/empty_dir.ts:33]: nit: Is this necessary?

**Comment by kt3k** [comment]
Looks good to me. Can you fix CI errors (run `deno fmt` and `deno lint`) and address nitpick comments?

**Comment by sanjaiyan-dev** [comment]
> Looks good to me. Can you fix CI errors (run `deno fmt` and `deno lint`) and address nitpick comments?

Okay 🫱🏼‍🫲🏾

**Comment by sanjaiyan-dev** [comment]
Done :)

File: fs/copy.ts [hunk 3]
```diff
@@ -190,17 +190,21 @@ async function copyDir(
   src = toPathString(src);
   dest = toPathString(dest);
 
+  const promisesCopyAsyncFunctions: Promise<void>[] = [];
```
Comment (iuioiua) [inline fs/copy.ts:193]:
```suggestion
  const promises = [];
```
I feel like this would've been sufficient.

File: http/cookie_map.ts [hunk 4]
```diff
@@ -607,9 +607,11 @@ export class SecureCookieMap extends CookieMapBase {
   /** Sets all cookies in the {@linkcode Request} to be deleted in the
    * response. */
   async clear(options: SecureCookieMapSetDeleteOptions) {
+    const promisesCookiesKeys: Promise<this>[] = [];
```
Comment (iuioiua) [inline http/cookie_map.ts:610]:
```suggestion
    const promises = [];
```
Ditto. This, I think, would've sufficed.

**Review by iuioiua — COMMENTED** [review]
Just a couple of nits.

**Comment by sanjaiyan-dev** [comment]
> Just a couple of nits.

Done 👍🏽

File: fs/copy.ts [hunk 5]
```diff
@@ -190,17 +190,21 @@ async function copyDir(
...
   dest = toPathString(dest);
 
+  const promises = [];
+
   for await (const entry of Deno.readDir(src)) {
     const srcPath = join(src, entry.name);
     const destPath = join(dest, basename(srcPath as string));
     if (entry.isSymlink) {
-      await copySymLink(srcPath, destPath, options);
+      promisesCopyAsyncFunctions.push(copySymLink(srcPath, destPath, options));
```
Comment (iuioiua) [inline fs/copy.ts:199]:
```suggestion
      promises.push(copySymLink(srcPath, destPath, options));
```
Sorry, I should've mentioned that these also need to be renamed.

File: fs/copy.ts [hunk 6]
```diff
@@ -190,17 +190,21 @@ async function copyDir(
...
     } else if (entry.isDirectory) {
-      await copyDir(srcPath, destPath, options);
+      promisesCopyAsyncFunctions.push(copyDir(srcPath, destPath, options));
```
Comment (iuioiua) [inline fs/copy.ts:201]:
```suggestion
      promises.push(copyDir(srcPath, destPath, options));
```
Ditto

File: fs/copy.ts [hunk 7]
```diff
@@ -190,17 +190,21 @@ async function copyDir(
...
     } else if (entry.isFile) {
-      await copyFile(srcPath, destPath, options);
+      promisesCopyAsyncFunctions.push(copyFile(srcPath, destPath, options));
```
Comment (iuioiua) [inline fs/copy.ts:203]:
```suggestion
      promises.push(copyFile(srcPath, destPath, options));
```

File: fs/empty_dir.ts (hunk 2 above)
Comment (iuioiua) [reply fs/empty_dir.ts:33]: Yeah, this doesn't seem necessary.

File: fs/copy.ts [hunk 8]
```diff
@@ -190,17 +190,21 @@ async function copyDir(
...
     }
   }
+
+  await Promise.all(promisesCopyAsyncFunctions);
```
Comment (iuioiua) [inline fs/copy.ts:207]:
```suggestion
  await Promise.all(promises);
```

File: http/cookie_map.ts [hunk 9]
```diff
@@ -607,9 +607,11 @@ export class SecureCookieMap extends CookieMapBase {
   /** Sets all cookies in the {@linkcode Request} to be deleted in the
    * response. */
   async clear(options: SecureCookieMapSetDeleteOptions) {
+    const promises = [];
     for await (const key of this.keys()) {
-      await this.set(key, null, options);
+      promisesCookiesKeys.push(this.set(key, null, options));
```
Comment (iuioiua) [inline http/cookie_map.ts:612]:
```suggestion
      promises.push(this.set(key, null, options));
```

File: http/cookie_map.ts [hunk 10]
```diff
@@ -607,9 +607,11 @@ export class SecureCookieMap extends CookieMapBase {
...
     }
+    await Promise.all(promisesCookiesKeys);
```
Comment (iuioiua) [inline http/cookie_map.ts:614]:
```suggestion
    await Promise.all(promises);
```

**Review by iuioiua — CHANGES_REQUESTED** [review]

**Comment by sanjaiyan-dev** [comment]
Sorry, and fixed it 🤝 @iuioiua

File: fs/empty_dir.ts (hunk 2 above)
Comment (iuioiua) [reply fs/empty_dir.ts:33]: @sanjaiyan-dev

**Review by iuioiua — APPROVED** [review]
LGTM 👍🏾

**Comment by sanjaiyan-dev** [comment]
> LGTM 👍🏾

Tks 🤝

**Comment by iuioiua** [comment]
Btw, I’d also change the title to something like “perf: use `Promise.all()`”

**Comment by sanjaiyan-dev** [comment]
> Btw, I’d also change the title to something like “perf: use `Promise.all()`”
@iuioiua 
Hope current title is in correct format 🤞🏽

**Review by kt3k — APPROVED** [review]
LGTM. Thanks!

**Comment by sanjaiyan-dev** [comment]
> LGTM. Thanks!

🤝 🤩

**Pull request merged**